api/src/main/java/brooklyn/basic/AbstractBrooklynObjectSpec.java

```diff
@@ -121,4 +121,21 @@ protected final void checkIsImplementation(Class<?> val, Class<? super T> requir
         if (Modifier.isAbstract(val.getModifiers())) throw new IllegalStateException("Implementation "+val+" is abstract, but must be a non-abstract class");
     }
 
+    @Override
+    public boolean equals(Object obj) {
+        if (obj==null) return false;
```

Comment:
You're right - it's not braking the official hashCode contract. It would just be wrong to put a spec in a `HashSet` and then modify it.

The spec is used like a builder, so tricky to make it immutable. We could have a `build` method (or `toImmutable` or some such), but I don't think it's a big deal.

Agree moving it to a test utility would be much harder to find (i.e. bad).

I'm happy to leave as-is in this PR.